content/blog/2025-12-16-allo-xls-backend.md

```diff
@@ -0,0 +1,218 @@
+
+## Challenges
+
+### The lowering passes
+For lowering Allo kernels into DSLX `fn`s, the biggest conceptual challenge was bridging the semantic gap between Allo's imperative memory model and XLS's functional purity constraint. Allo programs freely perform `affine.load` and `affine.store` operations on memref buffers, treating arrays as mutable storage. DSLX forbids in-place mutation. To bridge this gap, we translate each `affine.store` into a DSLX `update` expression. For a 2D store like `C[i][j] = value`, the lowering emits `update(C, i, update(C[i], j, value))`, constructing a new array rather than mutating in place.
+
+Meanwhile, the systolic array lowering is brittle in ways that feel unsatisfying from a compiler engineering perspective. We rely heavily on syntactic heuristics (e.g. substring matching on function and buffer names, regular expression parsing of stringified MLIR expressions, hardcoded assumptions about argument positions and nesting depths) rather than true semantic analysis. A robust solution would require either significantly more sophisticated pattern recognition using static analysis and dataflow analysis or a general IR transformation framework capable of inferring process networks from arbitrary imperative code. Both are substantially harder problems than we could tackle in this project. Our approach works for the systolic arrays Allo generates today, but it's fragile.
+
+### XLS Toolchain Quirks
+
+Getting systolic arrays to compile through XLS's full pipeline (DSLX → IR → optimized IR → pipelined Verilog) required substantial debugging. We had to systematically debug by isolating which XLS passes were failing, then examining the intermediate IR representations and optimizations applied.  This debugging process was time-consuming and required deep understanding of XLS's internal pipeline stages. Documentation was sparse for these advanced features, so we relied on reading XLS source code and experimenting with minimal test cases. Professor Sampson's pointer to the [XLS matmul_4x4 example](https://github.com/google/xls/blob/main/xls/examples/matmul_4x4/matmul_4x4.x) and the [proc tutorial](https://google.github.io/xls/tutorials/how_to_use_procs/#channel-arrays-and-loop-based-spawning) proved invaluable for understanding channel arrays and loop-based spawning patterns. In the midst of lowering the designs to Verilog, we found a bug in the default fifo configurations, which was fixed the next day! But, during the process of this project we found out that the golden matmul example couldn't properly lower to Verilog, as XLS's scheduling created circular dependencies within their fifo configuration. Thus, we had to create our own XLS systolic array example by playing around with DSLX and seeing what could actually get lowered to Verilog.
```

Comment:
It would be really really good to report this bug in the DSLX systolic array example to the XLS team! I'm sure they would appreciate it.